### PR TITLE
sds: refactor istio-agent SDS flow

### LINKED ISSUE #27691: VM mesh expansion, fails on restart
**Bug description**
Extending the mesh by one VM, setup done accordingly to [1.7 manual](https://istio.io/latest/docs/setup/install/virtual-machine/)

Mesh gets extended, VM shows up with **istioctl proxy-status**, fine so far.

```
nexus.default                                          SYNCED     SYNCED     SYNCED     SYNCED     istiod-6674ccc856-gknmq     1.7.0
```

#### sidecar.env was adjusted to read/write the certificates from/to a non volatile file path.
```
PROV_CERT=/etc/certs
OUTPUT_CERTS=/etc/certs
```

After a VM restart, the istio-sidecar fails to start because it expects an istio-token in /var/run/secrets/tokens/istio-token, which is obviously not there because of the restart and the volatile behaviour of /var/run.
I'm really confused, why this token is needed, when the certificates are available under /etc/certs, I was assuming the short lived token is only needed for initializing and getting the first set of certificates.

```
2020-10-02T12:10:08.321391Z     error   sds     Failed to get credential token: open ./var/run/secrets/tokens/istio-token: no such file or directory
2020-10-02T12:10:08.321453Z     info    sds     resource:default connection is terminated: rpc error: code = Canceled desc = context canceled
2020-10-02T12:10:30.771677Z     info    sds     resource:default new connection
2020-10-02T12:10:30.771737Z     error   sds     failed to get credential token: open ./var/run/secrets/tokens/istio-token: no such file or directory from path ./var/run/secrets/tokens/istio-token
2020-10-02T12:10:30.771744Z     error   sds     Failed to get credential token: open ./var/run/secrets/tokens/istio-token: no such file or directory
2020-10-02T12:10:30.771800Z     info    sds     resource:default connection is terminated: rpc error: code = Canceled desc = context canceled
2020-10-02T12:10:32.459164Z     info    sds     resource:default new connection
2020-10-02T12:10:32.459222Z     error   sds     failed to get credential token: open ./var/run/secrets/tokens/istio-token: no such file or directory from path ./var/run/secrets/tokens/istio-token
2020-10-02T12:10:32.459228Z     error   sds     Failed to get credential token: open ./var/run/secrets/tokens/istio-token: no such file or directory
2020-10-02T12:10:32.459308Z     info    sds     resource:default connection is terminated: rpc error: code = Canceled desc = context canceled
2020-10-02T12:11:01.539130Z     info    sds     resource:default new connection
2020-10-02T12:11:01.539203Z     error   sds     failed to get credential token: open ./var/run/secrets/tokens/istio-token: no such file or directory from path ./var/run/secrets/tokens/istio-token
2020-10-02T12:11:01.539211Z     error   sds     Failed to get credential token: open ./var/run/secrets/tokens/istio-token: no such file or directory
2020-10-02T12:11:01.539275Z     info    sds     resource:default connection is terminated: rpc error: code = Canceled desc = context canceled
```

In which real world setup you can expect a VM to not get restarted?
If the token is always needed I don't understand what the certificates are good for (except the root one)
What happens, without restart, but after the token has expired when it is constantly used?

[ ] Docs
[x ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ x] User Experience
[ ] Developer Infrastructure

**Expected behavior**
If valid certificates are existing in OUTPUT_CERTS or PROV_CERT they should get precedence and no token is needed.
If only the root cert is there, the short lived token is needed to seed the OUTPUT_CERTS with newly fetched certificates.

**Steps to reproduce the bug**
Restart a VM that is part of the mesh.

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**
```
client version: 1.7.3
control plane version: 1.7.3
data plane version: 1.7.0 (1 proxies), 1.7.3 (4 proxies)

Client Version: v1.19.2
Server Version: v1.16.13-gke.401
```

**How was Istio installed?**
```
istioctl install -f values.yaml
```

**Environment where bug was observed (cloud vendor, OS, etc)**
```
GCE, GKE
```

### LINKED ISSUE #28712: Refactor agent SDS handling
Now that gateway SDS will be removed, we should use this as an opportunity to refactor the SDS handling of the agent.

Goals:
* Reduce code and complexity
* Expand testing
* Improve performance, in particular startup latency

Current design:
* Custom XDS implementation, not following best practices we have learning in Istiod
* Secrets generated on demand, stored in cached designed for multi-tenant node agent and gateway secrets. Poll every x minutes to refresh certs. This is slow (we wait seconds to even start the CSR exchange flow) and complex because it was designed for other use cases. This also leads to bugs like https://github.com/istio/istio/issues/28099 and https://github.com/istio/istio/issues/28683.

Suggested design:
* Decouple fetching secrets from serving them.
     * Fetch on startup - will improve startup speed, make things simpler. Probably also makes the root cert handling a lot simpler
     * Change rotation to be event driven rather than polling. On update, trigger push like today
     * Expose some interface like FetchSecret(). If called on default/rootca, we return that, if its a file one we lookup the file, return it, and add a watch.
* Use Pilot's XDS implementation
  * This is a bit coupled to Istiod currently, but can be broken out with some minor changes

### PR DESCRIPTION
**Pull request by howardjohn**
This PR introduces some extensive refactoring of the entire SDS flow of the istio-agent.

This has two *functionality* changes as well:
1. Initial CSR will be sent in parallel to XDS request at start. In my testing, this reduces the proxy startup time by as much 30%.
1. For VMs, there is a circular dependency between XDS and SDS currently; we don't get any SDS requests until XDS request is received, but we need the SDS request in order to get the certificates written for mTLS auth to XDS server. This is worked around in 1.8 by using JWT for initial request; this will not work for systems that required exclusively mTLS authentication, such as when the VM is providing its own bootstrapped identity. This is fixed by the same change as above, by decoupling the initial CSR from the XDS flow.

Additionally, there are many refactorings. I strongly recommend reading through pkg/istio-agent/README.md first which describes the whole architecture now; I will mention only the differences here.

Previously, the SDS Server, Secret Cache, and ca client functionality sort of blended together, and all logic was duplicated between the 3. For example, the SDS server passes the cache a token, the cache does some processing on the token and sends it to the ca client. The cache handles exchanging and retrying of tokens. All of these makes the code more complex, inefficient, and duplicates logic. The new architecture has stricter distinction between the 3 packages. 

The SDS server is a "dumb" server acting as a bridge between SecretCache and Envoy, by turning SDS requests in GenerateSecret calls. This is re-written to use the same XDS server istiod uses, which has been battle tested for years. 

The SecretCache is mostly handling the same functionality, but refactored how it does this. Rather than caching per connection, like the old server did, the resource name is the only cache key. This simplifies things greatly, as well as allowing for the 2 functional changes mentioned above. The old connection key was from the legacy multi-tenant node-agent, so its not needed anymore. Additionally, we do not have a background job to check for certificate rotations (which can result in missed rotations, see https://github.com/istio/istio/issues/28683#issuecomment-723698368); instead we use a queue that will trigger an update at exactly the configured time to rotate the cert based on the TTL + grace period config.

All token logic has been moved into the ca client. This is done by a new TokenProvider struct, which implements the grpc.PerRPCCredentials interface; this same logic is used in the xds-proxy as well now. This also handles the token exchange and credential fetcher logic, so all of the logic is in a single place. Additionally, all retry logic, which previously was pushed into SecretCache, is now directly handled in the ca clients using `grpcretry` options.

In addition to the changes, high level "integration" tests are added for all the supported setups. This gives us coverage of flows that were previously uncovered, such as the implicit `/etc/certs` logic.

This PR has been split into a few PRs. As they are merged, the diff here will get smaller.
* [x] https://github.com/istio/istio/pull/29795
* [x] https://github.com/istio/istio/pull/29794
* [x] https://github.com/istio/istio/pull/29793
* [x] https://github.com/istio/istio/pull/29801
* [x] https://github.com/istio/istio/pull/29807
* [x] https://github.com/istio/istio/pull/29808

Fixes https://github.com/istio/istio/issues/28712
Fixes https://github.com/istio/istio/issues/27691

### CODE REVIEW

**Review by JimmyCYJ — COMMENTED** [review]
I just finished the review of pkg/istio-agent/README.md. Will review the rest and send out more comments.

File: pkg/istio-agent/README.md
Comment (howardjohn) [reply pkg/istio-agent/README.md:27]: Envoy hot restart is no longer used. Its still possible though, if a user disables mTLS then re-enables it I think we would end up with no SDS connections open while its disabled

File: pkg/istio-agent/docs/ca.dot
Comment (JimmyCYJ) [inline pkg/istio-agent/docs/ca.dot:1]: Are these ca.dot and ca.svg autogenerated?

File: pkg/security/security.go
Comment (JimmyCYJ) [inline pkg/security/security.go:91]: The original comment is incomplete, and is still incomplete.

File: pkg/security/security.go
Comment (JimmyCYJ) [inline pkg/security/security.go:135]: IIRC, this RotationInterval was 5 minutes, so that in the grace period, if the cert refreshing fails, it will re-refresh 5 minutes later. Do we have a similar logic that could handle the cert refresh scenario?

File: security/pkg/nodeagent/caclient/credentials.go
Comment (JimmyCYJ) [inline security/pkg/nodeagent/caclient/credentials.go:28]: Please add a short comment above TokenProvider.

File: security/pkg/nodeagent/caclient/credentials.go
Comment (JimmyCYJ) [inline security/pkg/nodeagent/caclient/credentials.go:65]: Why is this always false?

File: security/pkg/nodeagent/caclient/credentials.go
Comment (JimmyCYJ) [inline security/pkg/nodeagent/caclient/credentials.go:70]: Does `forCA` indicate whether this token is going to be added into CSR? When does t.forCA return false?

File: security/pkg/nodeagent/caclient/credentials.go
Comment (JimmyCYJ) [inline security/pkg/nodeagent/caclient/credentials.go:69]: Could you add a brief comment describing the logic of GetToken(), including how the token is fetched in every case?

**Review by JimmyCYJ — COMMENTED** [review]
I haven't finished yet, just send out some comments and continue to review the rest.

File: pilot/pkg/xds/ads_common.go
Comment (howardjohn) [reply pilot/pkg/xds/ads_common.go:26]: This is one of the last TODOs I had before marking this PR not WIP. This should be reverted

File: pkg/istio-agent/docs/ca.dot
Comment (howardjohn) [reply pkg/istio-agent/docs/ca.dot:1]: ca.dot is made by me, ca.svg is the rendered form for the markdown file

Comment (howardjohn) [reply pkg/istio-agent/docs/ca.dot:1]: basically just the source of the diagram. Its not required to keep it here but may be useful if we need to change it

File: pkg/security/security.go
Comment (howardjohn) [reply pkg/security/security.go:135]: Interesting, I hadn't considered the refresh failing. Let me see if I can add a test for this case. However, what I expect is:

* Cert refresh is triggered, we send an update to sds server.
* SDS server calls GenerateSecret, which fails. The XDS server returns error up to envoy, causing envoy to disconnect.
* Envoy will repeatedly reconnect to sds server and re-attempt GenerateSecret until the CA is back online

So the end result is we are resilient to CA downtime during refresh

File: security/pkg/nodeagent/caclient/credentials.go
Comment (howardjohn) [reply security/pkg/nodeagent/caclient/credentials.go:65]: Basically, if we set it to `true`, any requests using this without TLS will fail. Meaning JWT+plaintext is not possible. I wasn't really sure the right setting here... open to opinions

File: pilot/pkg/xds/ads_common.go
Comment (howardjohn) [reply pilot/pkg/xds/ads_common.go:26]: Most of the pilot/pkg/xds changes are broken out to https://github.com/istio/istio/pull/29794, so we won't need to review them here if that is merged there. Its a lot of boilerplate since a common interface needed to change

File: pilot/pkg/xds/ads.go
Comment (howardjohn) [reply pilot/pkg/xds/ads.go:138]: Previously, pilot/pkg/xds only handled ADS. With this change, we are also serving direct SDS, and in the future probably will serve ECDS as well, so I thought it made sense to move to the more generic name. But the log scope is still `ads` so maybe its not worth it

File: security/pkg/nodeagent/caclient/credentials.go
Comment (howardjohn) [reply security/pkg/nodeagent/caclient/credentials.go:70]: There is a long comment above that may explain this more, basically CA supports CredentialFetcher and TokenExchanger but XDS doesn't

File: pkg/security/security.go
Comment (howardjohn) [reply pkg/security/security.go:132]: I added a test case for this

File: security/pkg/nodeagent/caclient/credentials.go
Comment (JimmyCYJ) [reply security/pkg/nodeagent/caclient/credentials.go:29]: Still missing comment?

**Comment by JimmyCYJ** [comment]
Are you going to break remaining SDS change into more PRs? If yes, please let me know and I will take a look. At meantime, I will continue reviewing the rest.

**Comment by howardjohn** [comment]
The remaining bits will be hard to separate. If you have suggestions on parts to split to make it easier please let me know, if not I would just keep this as is

File: pilot/pkg/xds/ads.go
Comment (JimmyCYJ) [reply pilot/pkg/xds/ads.go:138]: Got it. This mix of ADS and XDS in logs may be a bit confusing. Would be better to unify the terms later.

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (JimmyCYJ) [inline security/pkg/nodeagent/sds/sdsservice.go:172]: Can we keep the implementation of FetchSecrets()?

**Review by JimmyCYJ — COMMENTED** [review]
Is is possible to split the secretmanager into a separate PR, and let @myidpt @shankgan @liminw to take a look? This is to make sure the leads/feature developers are all aware of this change, which would help code maintenance.

File: security/pkg/nodeagent/cache/secretcache.go
Comment (howardjohn) [reply security/pkg/nodeagent/cache/secretcache.go:122]: I don't see why not, we use it in other parts of the code and it has good test coverage

File: security/pkg/nodeagent/cache/secretcache.go
Comment (howardjohn) [reply security/pkg/nodeagent/cache/secretcache.go:151]: yeah we can probably make it a copy

File: security/pkg/nodeagent/cache/secretcache.go
Comment (howardjohn) [reply security/pkg/nodeagent/cache/secretcache.go:210]: The new logic cannot be blocked, unlikely the old one which would be blocked by envoy. This same logic is used in 10s of places in the pilot code base

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (howardjohn) [reply security/pkg/nodeagent/sds/sdsservice.go:172]: Why? It has not been used by envoy since xds v2 which has been a very long time

**Comment by howardjohn** [comment]
> Is is possible to split the secretmanager into a separate PR

The changes to sds and the secret manager are tightly coupled, the logic is very hard to split. Many of the simplifications on both sides are dependant on changes to one another

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (JimmyCYJ) [inline security/pkg/nodeagent/sds/sdsservice.go:65]: Could you briefly explain the logic in this ProxyNeedsPush and add comment here?

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (JimmyCYJ) [inline security/pkg/nodeagent/sds/sdsservice.go:139]: Could you please add a brief comment above Generate()? Which component is going to call Generate()?
Another question is, what does updates.Full mean?

File: security/pkg/nodeagent/sds/server.go
Comment (JimmyCYJ) [inline security/pkg/nodeagent/sds/server.go:52]: Comment above FillServer?

File: pkg/security/security.go
Comment (JimmyCYJ) [reply pkg/security/security.go:132]: By reading the changes in server.go and secretcache.go, the rotation task will clear the cache for `default`, and then call Server.UpdateCallback(`default`), which calls `Push` and adds a `PushRequest`. 
sdsservice.Generate()->sdsservice.generate()->SecretManagerClient.GenerateSecret(). If refresh fails, then pushes error to envoy and close the stream.

Does this new flow support multiple concurrent SDS streams? Do we have tests covering that?

**Review by JimmyCYJ — COMMENTED** [review]
The new flow LGTM overall. Thanks for this massive refactoring work. 

Once this PR is ready for a formal review, please add other reviewers in case I didn't catch any critical issues or didn't raise sufficient questions for clarification.

File: security/pkg/nodeagent/cache/secretcache.go
Comment (JimmyCYJ) [reply security/pkg/nodeagent/cache/secretcache.go:210]: That indefinite secret push is observed on a UDS socket, and root cause is not clear. Probably it does not happen on a TCP socket.

Comment (howardjohn) [reply security/pkg/nodeagent/cache/secretcache.go:210]: We are still using UDS, but the callback doesn't block. It just enqueues a new update, which the XDS server process asyncronously. For example, in the real Istiod XDS server, we trigger the same callback 100s of times per second for large scale clusters

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (howardjohn) [reply security/pkg/nodeagent/sds/sdsservice.go:139]: Added a comment. Basically, the xDS server dispatches the Envoy config generation to different implementators of Generate() based on TypeUrl.

For us, we have only one implementation. But for the main istiod xds, there are a bunch of generators for each type supported.

File: security/pkg/nodeagent/sds/server.go
Comment (howardjohn) [reply security/pkg/nodeagent/sds/server.go:52]: I removed the function, forgot to clean it up before

File: pkg/security/security.go
Comment (howardjohn) [reply pkg/security/security.go:132]: Yes, we do support this. That is one of the huge benefits of the change, we are use the battle-tested Istiod XDS server which has dealt with every obscure XDS edge case over the past few years. Also we have an explicit test for this in `TestSDS`

Comment (JimmyCYJ) [reply pkg/security/security.go:132]: Are there any README or docs describing the Istiod XDS server? That would help us to better understand and maintain the adoption of XDS server in other places outside Istiod.

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (JimmyCYJ) [reply security/pkg/nodeagent/sds/sdsservice.go:162]: Thanks for the information. And about the second question, by reading the comments in [type PushRequest struct {}](https://github.com/istio/istio/blob/6b967a0e36fa74bb2ffcf84a37b02b8bf5b415b2/pilot/pkg/model/push_context.go#L279), `// Full determines whether a full push is required or not. If false, an incremental update will be sent.` Is `updates.Full` always hardcoded to true in SDS use case?

File: pkg/security/security.go
Comment (howardjohn) [reply pkg/security/security.go:132]: Unfortunately no. I have been meaning to make one similar to the README in this PR, but haven't yet

File: pilot/pkg/xds/ads.go
Comment (costinm) [reply pilot/pkg/xds/ads.go:138]: If not too hard - I would revert this change. Who knows what tools for grepping and analyzing logs exist - not worth breaking them. But only if it's a simple replace ( will also reduce the change size )

File: pkg/istio-agent/README.md
Comment (costinm) [inline pkg/istio-agent/README.md:1]: Sounds like a good pattern, we should add similar docs !

File: pkg/istio-agent/README.md
Comment (costinm) [reply pkg/istio-agent/README.md:27]: Not clear if it's worth caching and returning the old cert, the system should be able to handle restarts of envoys when pods are rolled out.

Comment (howardjohn) [reply pkg/istio-agent/README.md:27]: I originally had the caching removed. The only reason I added it back is so to avoid 2x CSRs on startup (the "warming" one we send immediately, then in response to the SDS request). We also already have to cache the root cert (unless we send a full CSR on a request for it I suppose, which would be a bit odd/slow), so it wasn't too bad.

So for those 2 reasons I decided to add it back. I think the small complexity is worth reduced CSRs, unless you have a suggestion to have no cache and no repetitive CSRs?

File: pkg/istio-agent/README.md
Comment (costinm) [inline pkg/istio-agent/README.md:35]: Worth clarifying the 2 options:
- we can load existing certs created by a 3p tool, for example mounted Secrets or CSI - in this case we will NOT call Istiod to get certs, just use the file.
- for VM we may also use the on-disk cert to request a fresh cert - in this case OUTPUT_CERTS allows us to write down the new cert in the same place. There are 2 cases, where the provisioning cert is long-lived cert and we exchange it for a short-lived cert - in which case OUTPUT_CERTS will not be set or will point to a different dir, or cases where the original cert is also short lived and we need to rotate.

Finally, even if if JWT are the source - OUTPUT_CERTS is used to save the cert for other apps ( like prometheus, gRPC).

File: pkg/istio-agent/README.md
Comment (costinm) [inline pkg/istio-agent/README.md:48]: I think we still support the 'unscoped' JWT. Worth mentioning it.

File: pkg/istio-agent/agent.go
Comment (costinm) [inline pkg/istio-agent/agent.go:55]: Why ? We may still allow code ( like main ) to override those.

File: security/pkg/nodeagent/caclient/credentials.go
Comment (costinm) [reply security/pkg/nodeagent/caclient/credentials.go:69]: AFAIK the RFCs are very clear that JWTs MUST use secure transport. 

There may be a case in future for supporting IPsec/Wireguard - which is secure transport and may allow use of JWTs.

**Review by costinm — APPROVED** [review]
After first read - it looks great ! Left few comments, and I suspect after it's merged we should do additional tests and adjust if you missed any use case - but it's a clear improvement.

File: pkg/istio-agent/agent.go
Comment (howardjohn) [reply pkg/istio-agent/agent.go:55]: I think if these are modified they should be passed through security.Options rather than modifying global state, or it becomes harder to test/understand

File: pilot/pkg/xds/ads.go
Comment (howardjohn) [reply pilot/pkg/xds/ads.go:138]: Ok, I agree. reverted

File: pkg/istio-agent/README.md
Comment (shankgan) [inline pkg/istio-agent/README.md:77]: Might be worth clarifying these two points a further - 
(A) if PROV_CERT IS provisioned, then will this cert always be used to authenticate with control-plane servers (irrespective of whatever certs we get after CSR?)? Also, is there a possibility that the PROV_CERT is different for reaching the SDS server vs the XDS server?
(B)Can PROV_CERT and FILE_MOUNTED_CERTS be provisioned at the same time? My understanding is no. If the user provisions FILE_MOUNTED_CERTS, then these certs are meant to be used for all kinds of communication - for data-plane and control-plane communication

**Review by ericvn — APPROVED** [review]

File: pkg/istio-agent/README.md
Comment (howardjohn) [reply pkg/istio-agent/README.md:77]: A) is why I was suggesting we have the same config schema for xds and CA. Currently its somewhat arbitrary which fields apply to which server. I will document this discrepency

**Pull request merged**
